### ISSUE TITLE
Improve (by a lot) performance of layered navigation widgets

### ISSUE THREAD

**Issue by moory-se** [woothemes/woocommerce#31688]
Hi,

TL;DR
Adding an index of `(product_or_parent_id, term_id)` on `wp_wc_product_attributes_lookup` increases the performance of Layered Nav Widgets with between 9x and 550x.
TL;DR

The new lookup table (`wp_wc_product_attributes_lookup`) for sure have given a huge performance boost for **selecting the products**. However: the performance of the layered navigation widgets are, in some circumstances, still very slow. Thankfully, there's a super simple solution.

The WC_Widget_Layered_Nav is only shown when the chosen attribute actually is present in at least 1 of the products. Since sidebars (at least in a default setup) is global, you will have ALL your widgets present on every archive, even though they only will be shown on a subset of them. This means that if you have 50 Layered Navigation widgets in your sidebar, each one of them will have to issue a query to see if it should be shown or not.

This query is cached by the current url (or "request" - i'm not sure). This means that if a visitor actually applies a filter (let's say color:red), all widgets misses the cache and needs to recalculate its counts. This means that in practise, the "unfiltered" categories/archives will very often get cache hits, while "filtered" (such as ?filter_color=black) most often will not.

The query happens in `Automattic\W\I\P\Filterer::get_filtered_term_product_counts()`.

In our site, these queries are **very** slow. We have 97 attributes and 800 products. The ratio attribute:product might be high, but the actual numbers can't be described as extreme. The queries that is run looks something like this: 

```
SELECT COUNT(DISTINCT product_or_parent_id) AS term_count,
       term_id AS term_count_id
FROM wp_wc_product_attributes_lookup
LEFT JOIN wp_term_relationships ON (wp_wc_product_attributes_lookup.product_or_parent_id = wp_term_relationships.object_id)
INNER JOIN wp_posts ON wp_posts.ID = wp_wc_product_attributes_lookup.product_or_parent_id
WHERE wp_posts.post_type IN ('product')
  AND wp_posts.post_status = 'publish'
  AND (wp_wc_product_attributes_lookup.product_or_parent_id NOT IN
         (SELECT object_id
          FROM wp_term_relationships
          WHERE term_taxonomy_id IN (7) )
       AND wp_term_relationships.term_taxonomy_id IN (90, 91,  93, 94))
  AND wp_wc_product_attributes_lookup.taxonomy='pa_your-attribute-slug'
  AND wp_wc_product_attributes_lookup.term_id IN ('list', 'of', 'term', 'ids') 
GROUP BY wp_wc_product_attributes_lookup.term_id;
```

This query runs at 1.1 seconds. The more terms a given attribute has, the slower it gets. Please remember that this query runs for every widget.

I tried adding some indexes to the `wp_wc_product_attributes_lookup`. The most performant is what you would guess: `product_or_parent_id` and `term_id` (in that order). With that index in place, the slowest query goes from 1.1 seconds to 0.002 seconds. Running an `EXPLAIN` statement shows that the database has to examine 13 rows instead of 17341. It also obviously don't have to do the filesort (which is super slow), but instead a "index condition".

This is the statement:
```
ALTER TABLE `wp_wc_product_attributes_lookup`
ADD INDEX `product_or_parent_id_term_id` (`product_or_parent_id`, `term_id`);
```

So, by running the simple statement we decreased the loading time of the widget with ~550x. This is in the extreme cases (where there are lots of terms). In the best case (attributes with only 2 terms), the performance still got 9 times better (0.002 as opposed to 0.018).

I have not been able to measure the performance drop of writes. I've tried writing to the table and it's done in 0.0001 seconds. My database client can't show more decimals than that, so it's for sure negligible. Even if it was slower than this, it should not be a problem due to many reasons including:

- The writes are queued/deferred, so no-one will notice
- Frontend performance is obviously more important than backend performance
- This table is used for reading **a lot** more than for writing.

I wish I could PR this, but my non-existing understanding of database table changes in wordpress/woocommerce stops me from doing that. If someone reads this and get excited by making their site A LOT faster and knows how to create that PR, please do it and I'll be forever grateful. If you only get excited but don't know how to do it, please comment here and that might give the energy needed to understand how to actually create the PR :)

Thanks in advance,
Christoffer

**Comment by lanej0** [comment]
Hi @moory-se ,

Thank you for opening the issue! It requires further feedback from the WooCommerce Core team. I am adding the `needs developer feedback` label to this issue so that the Core team could take a look. 

Please note it may take a few days for them to get to this issue. Thank you for your patience.

**Comment by moory-se** [comment]
I noticed that in some cases, the query is longer. It contains some query for is_variation. The full query is pasted here. I will happily provide a database dump for testing, but I rather not paste it here in public.

[the-query.txt](https://github.com/woocommerce/woocommerce/files/7900925/the-query.txt)

I tried adding multiple indexes to see which ones the query planner would use, and found that we need 2 indexes to make this super fast.

**This is how the query looks without any indexes. Please note the "Rows" column.**

![image](https://user-images.githubusercontent.com/95882730/150223158-9fb389f6-7d9b-46be-a2fd-05fa2f74336a.png)

**This is how the query looks like with my initial proposal (1 index):**

![image](https://user-images.githubusercontent.com/95882730/150227024-9e9a2227-e8fd-4497-be46-7d55e72bccc1.png)

As you can see, the selected rows are still problematic. Let's fix it.

**This is how the most performant way looks:**

![image](https://user-images.githubusercontent.com/95882730/150223477-746cc663-119f-4ad2-a62c-1dbe4336d768.png)

Turns out that the most optimal (for reads), is to have 2 different indexes:
```
product_or_parent_id, term_id
is_variation_attribute, term_id
```

This particular query got from 0.0228 seconds (with 1 index) to 0.002 seconds (11,4x faster). For an attribute with more terms the query will be a lot slower without indexes, but virtually equally fast with indexes.

**Comment by vedanshujain** [comment]
hey @moory-se 

this sounds pretty neat, we would love to add this enhancement. if you are willing, can you also send us a PR given that you have already worked out a fix? otherwise we will be happy to take it up.

**Comment by moory-se** [comment]
Hi @vedanshujain,

I've implemented in our own sites, but I have not been able to do a PR. I simply don't know how to alter tables in upgrade scripts etc. Would be great if you could take it up ❤️

**Comment by Konamiman** [comment]
Pull request implementing this improvement as part of the database migration for 6.3: https://github.com/woocommerce/woocommerce/pull/31819
